### PR TITLE
Each process should get its own queue for IPC

### PR DESCRIPTION
**Pull request by usmanm**

### CODE REVIEW

File: src/backend/pipeline/cont_scheduler.c [hunk 1]
```diff
@@ -160,18 +165,18 @@ GetContQueryProcName(ContQueryProc *proc)
...
+			sprintf(buf, "combiner%d [%s]", proc->group_id, NameStr(proc->db_meta->db_name));
+			break;
+		case Worker:
+			sprintf(buf, "worker%d [%s]", proc->group_id, NameStr(proc->db_meta->db_name));
+			break;
+		case Adhoc:
+			sprintf(buf, "adhoc [%s]", NameStr(proc->db_meta->db_name));
+			break;
+		case Scheduler:
+			return "scheduler";
```
Comment (derekjn) [inline src/backend/pipeline/cont_scheduler.c]: Callers should expect to `pfree` whatever this function returns, so this should be `pstrduped`.

File: src/backend/pipeline/combinerReceiver.c [hunk 2]
```diff
@@ -81,91 +82,90 @@ combiner_receive(TupleTableSlot *slot, DestReceiver *self)
...
-		/* shard by groups */
-		tup->group_hash = hash_group(slot, c);
-	}
+		pts.hash = hash_group(slot, c);
 	else
-	{
-		/* shard by query id */
-		tup->group_hash = bms_singleton_member(c->queries);
-	}
+		pts.hash = MurmurHash3_64(&c->cont_executor->cur_query_id, sizeof(Oid), MURMUR_SEED);
```
Comment (derekjn) [inline src/backend/pipeline/combinerReceiver.c]: This should hash the query name instead of the id.

File: src/include/pipeline/dsm_cqueue.h [hunk 3]
```diff
@@ -42,24 +44,19 @@ typedef struct dsm_cqueue
...
+	atomic_ullong head;
+	atomic_ullong tail;
+	atomic_ullong cursor;
 
 	atomic_uintptr_t producer_latch;
 	atomic_uintptr_t consumer_latch;
 
-	dsm_cqueue_pop_fn pop_fn;
-	dsm_cqueue_cpy_fn cpy_fn;
+	dsm_cqueue_peek_fn peek_fn;
```
Comment (derekjn) [inline src/include/pipeline/dsm_cqueue.h]: Can you add some comments here or somewhere else that explain what each of these hooks is for?

File: src/backend/utils/misc/guc.c [hunk 4]
```diff
@@ -1523,7 +1523,7 @@ static struct config_bool ConfigureNamesBool[] =
 		 NULL,
 		},
 		&synchronous_stream_insert,
-		false,
+		true,
```
Comment (derekjn) [inline src/backend/utils/misc/guc.c]: Revert

File: src/test/regress/expected/pipeline_stats.out [hunk 5]
```diff
@@ -54,7 +54,7 @@ SELECT output_rows, errors, cv_create, cv_drop FROM pipeline_stats WHERE type =
 SELECT input_rows, input_bytes, errors FROM pipeline_stats WHERE type = 'worker';
  input_rows | input_bytes | errors 
 ------------+-------------+--------
-       2000 |      476000 |      0
+       2000 |      300000 |      0
```
Comment (derekjn) [inline src/test/regress/expected/pipeline_stats.out]: Nice!

File: src/backend/pipeline/cont_execute.c [hunk 6]
```diff
@@ -0,0 +1,715 @@
...
+}
+
+typedef struct ContQueryProcIPCState
+{
+	bool connected;
+	ContQueryProc *proc;
+	dsm_cqueue_handle *cq_handle;
+} ContQueryProcIPCState;
+
+static ContQueryProcIPCState *WorkerIPCStates = NULL;
```
Comment (derekjn) [inline src/backend/pipeline/cont_execute.c]: These and the `typdef` should go at the top of this file.

**Comment by derekjn** [comment]
Nice man, looks great! This is a huge, fundamental improvement. Once the tests are passing reliably go ahead and land it.

**Comment by usmanm** [comment]
TODO: Instead of having N segments per db, we have have one segment per DB of size N * ipc_mem and shard between all background procs. That way we don't have to connect and disconnect to a lot of segments. For adhoc, it stays the same.

**Comment by usmanm** [comment]
TODO: use CV name has instead of query id hash
TODO: move all state initialization logic to ContExecutor

**Pull request closed without merging**